js/src/viewer/manifestsPanel.js

```diff
@@ -129,15 +239,310 @@
+        clearManifestItems: function() {
+          this.manifestListItems = [];
+          this.manifestListElement.html('');
+        },
+        
+        // Set up 2-way correspondence between node IDs and URIs
+        registerNodeIdUriPair: function(nodeId, uri) {
+          // Map node ID to URI
+          this.nodeIdToUri[nodeId] = uri;
+          if (this.uriToNodeId[uri]) {
+            this.uriToNodeId[uri].push(nodeId); // Existing URI => new node ID
+          } else {
+            this.uriToNodeId[uri] = [nodeId]; // New URI => new node ID
+          }
+        },
+        
+        // Handler for selecting a new node
+        changeNode: function(node) {
+          var _this = this;
+          // Clean out manifest items on the right side
+          this.clearManifestItems();
+          // Listen in on manifests in the "folder" or collection it represents
+          switch (node.id) {
+            case 'preload': _this.expectedThings = _this.preloadedManifests; break;
+            case 'user': _this.expectedThings = _this.userManifests; break;
+            default: _this.expectedThings = _this.nodeManifests[node.id]; break;
+          }
+          // Populate and refresh the manifests listings
+          jQuery.each(_this.expectedThings, function(_, expectedThing) {
+            _this.addManifestFromUrl(expectedThing);
+          });
+          this.element.find('#manifest-search').keyup();
+          
+        },
+        
+        // Handler for expanding a node (> clicked)
+        expandNode: function(node) {
+          var _this = this;
+          // Update its children by loading their stated URIs
+          jQuery.each(_this.nodeCollections[node.id], function(_, uri) {
+            _this.updateCollectionFromUrl(uri, node.id);
+          });
+        },
+        
+        // Helper for loading a manifest or collection from a URL
+        addObjectFromUrl: function(url, source) {
+          var _this = this,
+            object = _this.state.getStateProperty('manifests')[url]; // Attempt to get from cache
+          // Cache hit
+          if (object) {
+            // Fire off the correct event if its cache entry is loaded
+            if (object.jsonLd) {
+              switch (object.jsonLd['@type']) {
+                case 'sc:Collection':
+                  _this.eventEmitter.publish('ADD_COLLECTION_FROM_URL', [url, source]);
+                break;
+                case 'sc:Manifest':
+                  _this.eventEmitter.publish('ADD_MANIFEST_FROM_URL', [url, source]);
+                break;
+              }
+            }
+          }
+          // Cache miss
+          else {
+            // Get the manifest or collection with AJAX
+            jQuery.ajax({
+              url: url,
+              dataType: 'json',
+              type: 'GET',
+              // Fire off the correct event once it finishes loading
+              success: function(data) {
+                switch (data['@type']) {
```

Comment:
Will do. Should I call it publishIiifData(data, url, source) and put it in utilities?